File: all-submissions/0329-longest-increasing-path-in-a-matrix/solution.java
class Solution {
    int m;
    int n;
    private static final int[][] directions = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    
    public int longestIncreasingPath(int[][] matrix) {
        if (matrix.length == 0) return 0;
        m = matrix.length;
        n = matrix[0].length;
        int[][] cache = new int[m][n];
        int ans = 0;
        for(int i=0; i< m;++i){
            for(int j=0; j<n;++j) {
                ans = Math.max(ans, dfs(i,j,matrix, cache));
            }
        }
        return ans;
    }

    public int dfs(int i, int j, int[][] matrix, int[][] cache){
        if(cache[i][j] != 0) return cache[i][j];
        for(int[] direction: directions) {
            int r= i+direction[0];
            int c = j+direction[1];
            if(r>=0 && c>=0 && r <m && c<n && matrix[r][c] > matrix[i][j]) {
                cache[i][j] = Math.max(cache[i][j], dfs(r, c, matrix, cache));
            }
        }
        return ++cache[i][j];
    } 

    public boolean isValid(int row, int col, int[][] matrix){
        return (row>=0 && col>=0 && row <m && col<n);
    }
}

File: all-submissions/0329-longest-increasing-path-in-a-matrix/solution.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        if len(matrix) == 0: return 0
        dir = [(0,1),(1,0),(0,-1),(-1,0)]
        m = len(matrix)
        n = len(matrix[0])
        dp = [[0]* n for _ in range(m)]
        ans = 0    
        
        def dfs(i, j):
            nonlocal m, n
            if dp[i][j] != 0: return dp[i][j]
            for r, c in dir:
                row = i+r
                col = j+c
                if row < 0 or row >= m or col < 0 or col >= n:
                    continue
                if matrix[row][col] > matrix[i][j]:
                    dp[i][j] = max(dp[i][j], dfs(row, col))
            dp[i][j]+= 1
            return dp[i][j]
       
        for i in range(m):
            for j in range(n):
                ans = max(ans, dfs(i, j))
        
        return ans


